tools/scoring/score/dimensions/dimension.py

```diff
@@ -113,8 +114,51 @@ def result_reporting(self) -> float:
             self.correct_ceiling_reporting
             ) if self.correct_ceiling_reporting else None
 
+  @staticmethod
+  def entity_is_canonical(*, entity: EntityInstance) -> bool:
+    """ Utility for determining whether an entity is canonical.
+    Used in "complex" dimensions to filter sets for comparison.
+
+    Args:
+      entity: An entity instance which has been appended
+      with a `type` attribute equal to its type from the universe
+
+    Returns:
+      Boolean indicating whether the entity's `type.is_canonical`
+    """
+    # NOTE: when passed to filter(), this will silently omit
+    # entities whose appended type is `None` (e.g. it was not found)
+    return getattr(entity.type, 'is_canonical', False)
+
+  @staticmethod
+  def entity_is_reporting(*, entity: EntityInstance) -> bool:
+    """
+    Utility for determining whether an entity is reporting.
+    Used in "complex" dimensions to filter sets for comparison.
+
+    Args:
+      entity: A standard entity instance
+
+    Returns:
+      Boolean indicating whether the entity has a `cloud_device_id`
+    """
+    return entity.cloud_device_id is not None
+
+  @staticmethod
+  def entity_is_virtual(*, entity: EntityInstance) -> bool:
```

Comment:
nit: `is_virtual_entity` ?

Reply:
just to be consistent with the other tooling on github

Reply:
(As above)